### PR TITLE
Add a hook for when AC caches a saved tensor

### PR DESCRIPTION
**Pull request by janeyx99**
Needed for https://github.com/pytorch/torchtune/pull/1913

Still need to add test

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139109

### CODE REVIEW

File: torch/utils/checkpoint.py [hunk 1]
```diff
@@ -1080,6 +1095,9 @@ def pack_hook(x):
                 _internal_assert(holder.handles.get(gid, None) is None)
                 holder.handles[gid] = _Handle()
                 target_frame.recomputed[gid][holder.handles[gid]] = x
+                global _global_checkpoint_saved_tensor_hooks
+                for hook in _global_checkpoint_saved_tensor_hooks.values():
+                    hook(x)
```
Comment (soulitzer) [inline torch/utils/checkpoint.py:1100]: Enforce that None is returned

File: torch/utils/checkpoint.py [hunk 2]
```diff
@@ -601,6 +602,20 @@ def _internal_assert(cond):
...
 
 
+_global_checkpoint_saved_tensor_hooks = OrderedDict()
+
+
+def _register_checkpoint_saved_tensor_hook(hook) -> RemovableHandle:
+    r"""Register a hook to be called when a tensor is saved by checkpoint.
+
+    The hook will be called with the tensor as the only argument.
+    """
```
Comment (soulitzer) [inline torch/utils/checkpoint.py:612]: add a comment to emphasize that this is a private API that people in general should not use

**Comment by soulitzer** [comment]
(from offline discussion) Maybe there are other ways of doing this that don't involve adding this hook, e.g. if we can check the use_count of the storage.

**Comment by janeyx99** [comment]
We don't need this right now, so not gonna add a new hook API

**Pull request closed without merging**